### DISCUSSION TITLE
Error relating to ADXL355 accelerometer library in micropyhon.

### DISCUSSION THREAD

**Discussion by jwpark98** [micropython/micropython#10271] [Libraries & Drivers]
I am new to micropython. Due to this forum, I have achieved several success. Thank you all for this help. However, I recently suffered from dealing with the accelerometer, adxl355. As a mechanical engineer, I am very interested in measuring acceleration of an object. To do, I have tested several sensors such as adxl345, mpu6050, and adxl355. I completed adxl345 and mpu6050, but adxl355 is not easy due to its complexity. Below is my adxl355.py (mycropython api version) referred from adxl355.py (python version) and adxl345.py (micropython version). Especially, the code treating byte operator is not easy to me. I do not know how to deal with byte sections. Could you please let me know any good adxl355.py (micropython) or teach me how to modify this code. I could not find any reliable code for adxl355 in micropython.

---------- code ------------
# Import necessary libraries
from micropython import const
from machine import SPI, Pin
import time

# ADXL355 addresses
DEVID_AD = const(0x00)
DEVID_MST = const(0x01)
PARTID = const(0x02)
REVID = const(0x03)
STATUS = const(0x04)
FIFO_ENTRIES = const(0x05)
TEMP2 = const(0x06)
TEMP1 = const(0x07)

XDATA3 = const(0x08)
XDATA2 = const(0x09)
XDATA1 = const(0x0A)
YDATA3 = const(0x0B)
YDATA2 = const(0x0C)
YDATA1 = const(0x0D)
ZDATA3 = const(0x0E)
ZDATA2 = const(0x0F)
ZDATA1 = const(0x10)

FIFO_DATA = const(0x11)
OFFSET_X_H = const(0x1E)
OFFSET_X_L = const(0x1F)
OFFSET_Y_H = const(0x20)
OFFSET_Y_L = const(0x21)
OFFSET_Z_H = const(0x22)
OFFSET_Z_L = const(0x23)
ACT_EN = const(0x24)
ACT_THRESH_H = const(0x25)
ACT_THRESH_L = const(0x26)
ACT_COUNT = const(0x27)
FILTER = const(0x28)
FIFO_SAMPLES = const(0x29)
INT_MAP = const(0x2A)
SYNC = const(0x2B)
RANGE = const(0x2C)
POWER_CTL = const(0x2D)
SELF_TEST = const(0x2E)
RESET = const(0x2F)

# Data ranges
RANGE_2G = const(0x01)
RANGE_4G = const(0x02)
RANGE_8G = const(0x03)

# Values
READ_BIT = 0x01
WRITE_BIT = 0x00
DUMMY_BYTE = 0xAA

class ADXL355:
   def __init__(self, spi_id, cs_pin=5, scl_pin=18, sda_pin=23, sdo_pin=19, spi_freq=10000000):
   #SPI0, id=0 / SPI1(HSPI), id=1 / SPI2(VSPI), id=2
   # SPI pins
      self.spi = SPI(spi_id, sck=Pin(scl_pin, Pin.OUT), mosi=Pin(sda_pin, Pin.OUT), miso=Pin(sdo_pin), baudrate=spi_freq, polarity=1, phase=1, bits=8, firstbit=SPI.MSB)
      time.sleep(0.2)
      self.cs = Pin(cs_pin, Pin.OUT, value=1)
      time.sleep(0.2)

   def __del__(self):
      self.spi.deinit()

   # == writing ==
   def write_data(self, regaddr:int, the_byte:int):
      self.cs.value(0)
      self.spi.write(bytearray((regaddr, the_byte)))
      self.cs.value(1)
      return self

   # == readings ==
   def read_data(self, regaddr: int, nbytes: int) -> bytearray or int:
      self.cs.value(0)
      value = self.spi.read(nbytes + 1, regaddr)[1:]
      self.cs.value(1)
      return value
        
   #buf = bytearray(4)
   def read_into(self, buf: bytearray, regaddr: int) -> bytearray:
      self.cs.value(0)
      self.spi.readinto(buf, regaddr)
      self.cs.value(1)
      return buf

   def read_multiple_data(self, address_list):
      spi_ops = []
      for address in address_list:
         spi_ops.append(address << 1 | READ_BIT)
         spi_ops.append(DUMMY_BYTE)
         
      return self.spi.read(spi_ops)[1:]

   def get_temperature(self):
      # Reading data
      temp_data = [self.spi.read(TEMP1, 4), self.spi.read(TEMP2, 4)]
      # Join data
      temp_data = (temp_data[0]) + ((temp_data[1] - ((temp_data[1] >> 4) << 4)) << 8)
      # temp_data = (1852 - temp_data) / 9.05 + 19.21
      return temp_data

   def get_acceleration(self):
      ## reading data
      xData = [self.read_data(XDATA1, 4), self.read_data(XDATA2, 4), self.read_data(XDATA3, 4)]
      yData = [self.read_data(YDATA1, 4), self.read_data(YDATA2, 4), self.read_data(YDATA3, 4)]
      zData = [self.read_data(ZDATA1, 4), self.read_data(ZDATA2, 4), self.read_data(ZDATA3, 4)]

      # Join data
      x_data = (xData[0] >> 4) + (xData[1] << 4) + (xData[2] << 12)
      y_data = (yData[0] >> 4) + (yData[1] << 4) + (yData[2] << 12)
      z_data = (zData[0] >> 4) + (zData[1] << 4) + (zData[2] << 12)

      # Apply two complement
      if x_data & 0x80000 == 0x80000:
         x_data = ~x_data + 1

      if y_data & 0x80000 == 0x80000:
         y_data = ~y_data + 1

      if z_data & 0x80000 == 0x80000:
         z_data = ~z_data + 1

      # Return values
      return [x_data, y_data, z_data]

   def _reset_settings(self):
        ## reset all settings
        self.write_data(RESET, 0x52)
    
   # == Get measuring range ==
   def _set_measure_range(self, measure_range:int):
       self.write_data(RANGE, measure_range)
       return self
        
if __name__ == '__main__':
   spi0 = ADXL355(spi_id=2, cs_pin=5, scl_pin=18, sda_pin=23, sdo_pin=19, spi_freq=10000000)
   ## Set sensor
   spi0._reset_settings()
   spi0._set_measure_range(RANGE_2G)
   
   f_s = 3 # Hz
   dt = 1/f_s # sec.

   axes = spi0.get_acceleration()     # pylint: disable=invalid-name
   print(axes)

---------------------------------------------

When I run the code, following error is occurred.

typeerror: unsupported types for __rshift__: 'bytes', 'int'

I think the error is related to the part as below:

x_data = (xData[0] >> 4) + (xData[1] << 4) + (xData[2] << 12)

Maybe I did some mistake to port adxl355.py (python version) into micropython version.

This work is pretty important to me since I am studying some measurement system to obtain acceleration data on the surface of machinery. Since I am not expert in micropython programming, I do not know where I have to start to modify my ugly code.

JW

**Comment by robert-hh** [comment]
xData is a list of bytearray objects, so xData[0] is a bytearray object, not an int. You may have intended e.g. xData[0][0] etc., which is an int, or convert the content of what has been returned by read_data89 to an int first and add that to the list.

**Reply by jwpark98** [reply]
Thanks for this quick reply. The function ‘read_data’ returns byte. Python code uses << or >> operator to get values however there is no example to treat them in micropython. Could you please show me how to get values as in same manner?

**Reply by jwpark98** [reply]
Do you mean that I needed to cast the returned type of read_data() as below? 
FROM
x_data = (xData[0] >> 4) + (xData[1] << 4) + (xData[2] << 12)
TO
x_data = (int(xData[0]) >> 4) + (int(xData[1]) << 4) + (int(xData[2]) << 12)

**Reply by jwpark98** [reply]
This also returned error messege.

- ValueError: invalid syntax for integer with base 10

**Reply by jwpark98** [reply]
When I print xData using self.read_Data(xData, it gave me [b’\xff\xff\xff\xff’, b’\xff\xff\xff\xff, b’\xff\xff\xff\xff]. I think it is something wrong.

**Reply by robert-hh** [reply]
The function read_data() returns a bytearray of five bytes, as your print shows, not a single integer. That why you get the coding error. And it returns five bytes because your code requests five bytes. You can convert that into an integer using e.g. struct.unpack().
The fact that this bytearray is all 0xff is a different problem. It seems that you do not get any data from the sensor. So check the wiring and the way you intialize the sensor.

**Comment by brokenbrigde9000** [comment]
@jwpark98 Did you manage to get the ADXL355 running with micropython?
